packages/repository/src/repositories/relation.repository.ts

```diff
@@ -0,0 +1,173 @@
+import {
+  DefaultCrudRepository,
+  DataSourceType,
+  EntityCrudRepository,
+  EntityRepository,
+  CrudRepository,
+  Repository,
+  juggler,
+} from '.';
+import {
+  Entity,
+  Filter,
+  AnyObject,
+  Where,
+  DataObject,
+  Options,
+  FilterBuilder,
+  WhereBuilder,
+} from '..';
+
+/**
+ * CRUD operations for a target repository of a HasMany relation
+ */
+export interface HasManyEntityCrudRepository<T extends Entity, ID>
+  extends Repository<T> {
+  /**
+   * Create a target model instance
+   * @param targetModelData The target model data
+   * @param options Options for the operation
+   * @returns A promise which resolves to the newly created target model instance
+   */
+  create(targetModelData: DataObject<T>, options?: Options): Promise<T>;
+
+  /**
+   * Build a target model instance
+   * @param targetModelData The target model data
+   * @returns A promise of the created model instance
+   */
+  build(targetModelData: DataObject<T>): Promise<T>;
+
+  /**
+   * Find a related entity by id
+   * @param id The foreign key
+   * @param options Options for the operation
+   * @returns A promise of an entity found for the id
+   */
+  findById(id: ID, filter?: Filter, options?: Options): Promise<T>;
+
+  /**
+   * Update a related entity by foreign key
+   * @param data Data attributes to be updated
+   * @param id Value for the foreign key
+   * @param options Options for the operation
+   * @returns Promise<true> if the entity is updated, otherwise
+   * Promise<false>
+   */
+  updateById(id: ID, data: DataObject<T>, options?: Options): Promise<boolean>;
+
+  /**
+   * Delete a related entity by id
+   * @param id Value for the entity foreign key
+   * @param options Options for the operation
+   * @returns Promise<true> if an entity is deleted for the id, otherwise
+   * Promise<false>
+   */
+  deleteById(id: ID, options?: Options): Promise<boolean>;
+
+  /**
+   * Check if the related entity exists for the given foreign key
+   * @param id Value for the entity foreign key
+   * @param options Options for the operation
+   * @returns Promise<true> if an entity exists for the id, otherwise
+   * Promise<false>
+   */
+  exists(id: ID, options?: Options): Promise<boolean>;
+}
+
+export class DefaultHasManyEntityCrudRepository<T extends Entity, ID>
+  implements HasManyEntityCrudRepository<T, ID> {
+  /**
+   * Constructor of DefaultCrudRepository
+   * @param targetRepo a repository of the target model
+   * @param foreignKey the constraints to scope target repo CRUD methods
+   */
+  constructor(
+    public targetRepo: DefaultCrudRepository<T, ID>,
+    public primaryKey: ID,
+  ) {}
+  execute(
```

Comment:
IMO `execute` does not belong to a scoped repository. How do you envision to implement scoping of an arbitrary SQL command?

I think this may be a sign that `execute` does not belong to the `Repository` interface and that we should have another interface for repositories allowing consumers to execute arbitrary command.

Reply:
I agree it shouldn't belong in this scoped repository or the `Repository` interface. Raymond and I were discussing the possibility of making `Repository` more granular by classifying it based on the type of persistence function i.e. `LookupRepository` would have all the `find` related CRUD methods, `WriteRepository` (I'm sure there is a better name), would have the `create` related methods, `MutationRepository` might have the `update` and related methods (this might fall under the previous one, and `DestroyRepository` for deletes. Thoughts? IMO `execute` would probably just have its own repository since it's arbitrary SQL command(s).